### DISCUSSION TITLE
CPU memory leak when using PL on GCP TPUs

### DISCUSSION THREAD

**Discussion by Natithan** [Lightning-AI/pytorch-lightning#7702] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi,

I've been stuck for quite some time on this.
I am training a VilBERT-like model, and because each training run takes quite a long time, I am running it on google cloud TPUs in the hope of speeding it up.

It doesn't work however, as during training the CPU memory slowly increases until it's killed by the VMs OOM-process-killer.
The TPU memory is not filling up however. (See image below).
I have 64GB CPU RAM available, and am using a v3-8 TPU.
As I didn't observe this increase in CPU memory while training on the GPUs in my lab, I think it might have to do with how I am using the TPUs.
I'm using pytorch-lightning to move my code to TPUs, hence why I'm hoping someone on this forum can help me :).

I also checked the memory usage when using _one_ TPU core (rather than 8), and there too the memory increases per step (see image).
For that run, I also tracked memory usage using mprof. (see image)
What is odd, is that this gave me a really different graph however, one in which there is no leak apparently. I feel like this should give me some hint, but I'm not sure what :).

I've also tried to use memory_profiler to see the memory-changes line-by-line.
Because the program is multiprocess however, this gives quite a bit of seemingly unrelated memory fluctuation.

I did a bit of googling around, here are some related threads that didn't solve my problem:

- https://github.com/huggingface/transformers/issues/5509: this person didn't really fix their memory leak, but mitigated it by reducing how fast it leaked: they used a redis server to not load data in each subprocess.
- https://stackoverflow.com/questions/65107933/pytorch-model-training-cpu-memory-leak-issue in this thread, the issue was not detaching a tensor, causing it to keep the whole computation graph. As far as I can tell however, my code doesn't keep any tensors between steps that could 'build up' in this way.
- https://github.com/PyTorchLightning/pytorch-lightning/issues/1510 this issue seems related as well, but it should be solved in pytorch-lightning according to the thread.

If anyone had a similar problem, and could point me in the right direction (or even how to best find the cause), that would be great!

![image](https://user-images.githubusercontent.com/17345374/119503894-ba51b780-bd6b-11eb-84dc-6a8678a4e74e.png)
![image](https://user-images.githubusercontent.com/17345374/119505539-5af4a700-bd6d-11eb-8df5-dc7b568afcb9.png)
![image](https://user-images.githubusercontent.com/17345374/119505562-5fb95b00-bd6d-11eb-950d-ba5336a4a877.png)

**Comment by kaushikb11** [comment]
Hi @Natithan, thanks for posting this, will give it a look early next week. Did you give the new [TPU VMs](https://cloud.google.com/tpu/docs/users-guide-tpu-vm) a try? And let me know if the problem still persists?

**Reply by Natithan** [reply]
Hi!
I recently spotted the TPU VMs, it seems they have around 330GB of RAM so that might just be enough to mitigate my memory leak yes.
I haven't been able to connect them with an external persistent disk though: they have 100GB of disk by default, but my data is around 1300GB.
For now, I'm working my way around it by making sure I can resume from checkpoints mid-epoch (storing index into my data etc). But the leak itself isn't resolved yet.

Thanks!

Nathan

**Reply by kaushikb11** [reply]
For sure. @Natithan Could you confirmed the memory leak isn't an issue while training on GPUs or CPUs?

**Reply by Natithan** [reply]
Hey @kaushikb11 ,
I'm currently just running it on GPUs and will be busy trying to finish my paper for the next month or so.
I'll check the memory usage on GPU after I'm done with that :)

**Reply by Natithan** [reply]
Just to follow up on this, I won't be investigating this more in the foreseeable future as it's no longer a blocking issue for me (moved on to next project). Thanks for the help in any case!

**Reply by tchaton** [reply]
Dear @Natithan,

Any chance you could just check this out using the Lightning master ? We resolved several memory leaks issues and would appreciated if you could confirm it still exists.

Best,
T.C

**Reply by Natithan** [reply]
Hi @tchaton , it will be tricky to do so, as I already deleted the VMs on GCP: I ended up doing everything on GPUs locally. Sorry for that!
